### ISSUE TITLE
Atom reordering and HR matrix

### ISSUE THREAD

**Issue by marcelclaro** [deepmodeling/abacus-develop#3064]
### Details

Hi,
I am interested on use abacus to find tight-binding hamiltonians and hopping coefficients.
It seems the H(R) matrix output will work great for my propose, however, I am having troubles to find the right value. It seems that the the atom order and/or lattice vectors change internally (in the symmetry analysis maybe?) even if I use symmetry=0. After that I loose track of the atom, orbital and position.

My structure file is
ATOMIC_POSITIONS
Direct                  #Direct (coordinates of atom positions below would in fraction coordinates)
O
0.0
2
0.50000 0.50000 0.50000 m 1 1 1
0.00000 0.00000 0.50000 m 1 1 1
Pb
0.0
2
0.50000 0.00000 0.26207 m 1 1 1
0.00000 0.50000 0.73792 m 1 1 1

Comparing with the wannier_hr matrix, I found that apparently, SCF change the position of 0.00000 0.50000 0.73792  to 0.50000 0.00000 -0.26207 or so. Some expected values that should be found in H(0,0,0) are found in H(0,0,-1), etc...

There is any way to keep the exact same order written in the STRU file?

Thank you,
Marcel

### Have you read FAQ on the online manual http://abacus.deepmodeling.com/en/latest/community/faq.html

- [X] Yes, I have read the FAQ part on online manual.

### Task list for Issue attackers

- [X] Understand the problem or question described by the user.
- [X] Check if the issue is a known problem or has been addressed in the documentation.
- [ ] Test the issue or problem on a similar system or environment, if possible.
- [X] Identify the root cause or provide clarification on the user's question.
- [ ] Provide a step-by-step guide, including any necessary resources, to resolve the issue or answer the question.
- [X] If the issue is related to documentation, update the documentation to prevent future confusion (optional).
- [X] If the issue is related to code, consider implementing a fix or improvement (optional).
- [ ] Review and incorporate any relevant feedback from users or developers.
- [ ] Ensure the user's issue is resolved or their question is answered and close the ticket.

**Comment by Liu-RX** [comment]
Hi marcel,
Sorry for the late reply. I was trying to reproduce your self-consistent calculation. I tried Silicon atoms in (0.5, 0, 0.26207) and (0, 0.5, 0.73792) coordinates, and the atomic positions remain unchanged. So could you please send us your input files (INPUT, STRU, and KPT)?

**Comment by marcelclaro** [comment]
Hi Liu,
Sorry for the late reply.
I was busy and want to recheck everything before reply.

I debugged the ABACUS code. Mainly the list of index. In fact it seems the order was kept until the end and it is not a symmetry bug.
However the problem was the H(R) itself. I checked Hmn(R) and there are not equal to Hnm*(-R). I am missing something?


INPUT_PARAMETERS
suffix                  PbO_alpha
ntype                   2
pseudo_dir              /home/marcel/pseudos/SG15_ONCV_v1
orbital_dir		        /home/marcel/pseudos/SG15_StandardOrb
latname                 so
ecutwfc                 100             # Rydberg
scf_thr                 1e-9		# Rydberg
basis_type              lcao
symmetry            0            
calculation             scf	# this is the key parameter telling abacus to do a optimization calculation
vdw_method          d3_0
out_chg             1
out_mat_hs2  1

K_POINTS
0 
Gamma
8 8 8 0 0 0

#This is the atom file containing all the information
#about the lattice structure.

ATOMIC_SPECIES
O 15.99900 O_ONCV_PBE-1.0.upf
Pb 207.20000 Pb_ONCV_PBE-1.0.upf

NUMERICAL_ORBITAL
O_gga_7au_100Ry_2s2p1d.orb
Pb_gga_7au_100Ry_2s2p2d1f.orb

LATTICE_CONSTANT
7.72265  # Lattice constant in Angstroms

LATTICE_PARAMETERS 1.00000 1.22358


ATOMIC_POSITIONS
Direct                  #Direct (coordinates of atom positions below would in fraction coordinates)
O
0.0
2
0.50000 0.50000 0.50000 m 1 1 1
0.00000 0.00000 0.50000 m 1 1 1
Pb
0.0
2
0.50000 0.00000 0.26207 m 1 1 1
0.00000 0.50000 0.73793 m 1 1 1

**Comment by Liu-RX** [comment]
Hi Marcel,
Thank you for your response! I tried out your calculation with ABACUS, output the electronic density, and plot the density with VESTA. From the density distribution, it seems that the coordination of the second Pb atom remains (0.00000 0.50000 0.73793) unchanged. You could also do that with the `SPIN1_CHG.cube` file generated in the `OUT.*` folder. Could you please further explain how you find the atomic coordinates of the second Pb atom is incorrect?

<img width="677" alt="截屏2023-11-24 下午11 42 53" src="https://github.com/deepmodeling/abacus-develop/assets/75369672/d214b255-e166-4d80-b101-4df262a17bcf">

**Comment by marcelclaro** [comment]
Dear Liu,
Thank you for your answer.

Sorry, I did not explain quite well. The wavefunction and so on looks fine.
The only problem is with the HR ans SR matrices. Using "out_mat_hs2 1"

Initially I though it was a problem due to geometry, but actually it is a problem that the hamiltonian do not follow the property:
Hmn(R) = Hnm*(-R) 
So, depending on which Rs are chosen, I can get (or not) the right values, which are comparable to wannier90 hr file.

**Comment by Liu-RX** [comment]
> Dear Liu, Thank you for your answer.
> 
> Sorry, I did not explain quite well. The wavefunction and so on looks fine. The only problem is with the HR ans SR matrices. Using "out_mat_hs2 1"
> 
> Initially I though it was a problem due to geometry, but actually it is a problem that the hamiltonian do not follow the property: Hmn(R) = Hnm*(-R) So, depending on which Rs are chosen, I can get (or not) the right values, which are comparable to wannier90 hr file.

Dear Marcel,
Now I understand what you mean. I wrongly thought that if the charge density, wave functions etc are in order, then the H(R) is too. Sorry for that. I will check again.

**Comment by maki49** [comment]
What's the version of your ABACUS?
I tested your case with the newest develop branch, and found that H(0,0,1)(0,0) is equal to H(0,0,-1)(0,0), and no elements of other R are equal to H(0,0,0)(0,0). 
So I guess your problem has already been fixed.

(If the problem still exists in the newest develop branch, could you please point out which (m, n, R) does not satisfy Hmn(R)=Hnm*(-R)? It will help us a lot to locate the problem. Thanks. )

**Comment by Liu-RX** [comment]
Thanks for @maki49 's useful comment!
I was also able to plot the H(R) vs H(-R) from the `data-HR-sparse_SPIN0.csr` file for `R=(1, 0, 0), (0, 1, 0), (0, 0, 1)`. It seems that they satisfy the Hmn(R)=Hnm(-R) condition. I show in the following the case of `R=(1,0,0)`.
I also adopted the latest code from the repo. Maybe the bug has been fixed recently and the latest version works?

<img width="1204" alt="截屏2023-11-28 下午5 02 11" src="https://github.com/deepmodeling/abacus-develop/assets/75369672/265b1a37-abe5-47e5-82ed-8def6e49371f">
